src/ioctls/sev.rs

```diff
@@ -0,0 +1,680 @@
+        if ret == 0 {
+            Ok(data)
+        } else {
+            Err(errno::Error::last())
+        }
+    }
+}
+
+impl SevLaunch<Uninit> {
+    /// Instantiate a new SEV launch context with VmFd.
+    pub fn new_with_vm(vm: VmFd) -> Result<Self> {
+        let sev = OpenOptions::new().read(true).write(true).open("/dev/sev")?;
+
+        // Safe because we are the owners of the fd.
+        Ok(SevLaunch {
+            sev: new_sevfd(sev),
+            vm: vm,
+            state: Uninit,
+        })
+    }
+
+    /// Initialize the SEV platform context.
+    /// In a typical workflow, this command should be the first command issued.
+    pub fn init(self) -> Result<SevLaunch<Init>> {
+        self.sev_ioctl(SevCmd::Init, ())?;
+
+        Ok(SevLaunch {
+            sev: self.sev,
+            vm: self.vm,
+            state: Init,
+        })
+    }
+}
+
+impl From<SevLaunch<Uninit>> for SevLaunch<Init> {
+    fn from(launch: SevLaunch<Uninit>) -> SevLaunch<Init> {
+        launch
+            .init()
+            .expect("Launch state transition failed from Uninit to Init")
+    }
+}
+
+impl SevLaunch<Init> {
+    /// Create the memory encryption context according to user-provided `Start` information,
+    /// incl. a guest policy the owner's public Diffie-Hellman (PDH) key and session information.
+    pub fn start(self, start: Start) -> Result<SevLaunch<LaunchUpdate>> {
+        let start = kvm_sev_launch_start {
+            handle: 0,
+            policy: start.policy.into(),
+            dh_uaddr: &start.cert as *const _ as u64,
+            dh_len: size_of_val(&start.cert) as u32,
+            session_uaddr: &start.session as *const _ as u64,
+            session_len: size_of_val(&start.session) as u32,
+        };
+
+        let res = self.sev_ioctl(SevCmd::LaunchStart, start)?;
+
+        Ok(SevLaunch {
+            sev: self.sev,
+            vm: self.vm,
+            state: LaunchUpdate(Handle(res.handle)),
+        })
+    }
+}
+
+impl SevLaunch<LaunchUpdate> {
+    /// Encrypts a memory region specified by user in place.
+    /// It also calculates a measurement of the memory contents which is a signature
+    /// of the memory contents that can be sent to the guest owner as an attestation
+    /// that the memory was encrypted correctly by the firmware.
+    pub fn update_data(&self, data: &[u8]) -> Result<()> {
+        let update = kvm_sev_launch_update_data {
+            uaddr: data.as_ptr() as u64,
+            len: data.len() as u32,
+        };
+
+        self.sev_ioctl(SevCmd::LaunchUpdateData, update)?;
+
+        Ok(())
+    }
+
+    /// Retrieve the measurement of the data encrypted.
+    /// The guest owner may wait to provide the guest with confidential information until
+    /// it can verify the measurement thru comparison.
+    pub fn measure(self) -> Result<SevLaunch<LaunchSecret>> {
+        let mut mr = Measurement::default();
```

Comment:
This doesn't compile on my machine because there's no `impl Default` for `Measurement`.